overlord/snapstate/policy/policy.go

```diff
@@ -0,0 +1,184 @@
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+)
+
+func init() {
+	snapstate.PolicyFor = For
+}
+
+func For(typ snap.Type, model *asserts.Model) snapstate.Policy {
+	switch typ {
+	case snap.TypeKernel:
+		return &kernelPolicy{modelName: model.Kernel()}
+	case snap.TypeGadget:
+		return gadgetPolicy{}
+	case snap.TypeOS:
+		return &osPolicy{modelName: model.Base()}
+	case snap.TypeBase:
+		return basePolicy{}
+	default:
+		return appPolicy{}
+	}
+}
+
+// TODO: split this out into one file per type, leave only For and
+// helpers (and maybe snap.TypeApp impl) here.
+
+type appPolicy struct{}
+
+func (appPolicy) CanRemove(_ *state.State, snapst *snapstate.SnapState, all bool) bool {
+	if !all {
+		return true
+	}
+
+	return !snapst.Required
+}
+
+type gadgetPolicy struct{}
+
+func (gadgetPolicy) CanRemove(_ *state.State, _ *snapstate.SnapState, all bool) bool {
+	return !all
+}
+
+type kernelPolicy struct {
+	modelName string
+}
+
+func (p *kernelPolicy) CanRemove(_ *state.State, snapst *snapstate.SnapState, all bool) bool {
+	name := snapst.InstanceName()
+	if name == "" {
+		// not installed, or something. What are you even trying to do.
+		return false
+	}
+
+	if boot.InUse(name, snapst.Current) {
+		return false
+	}
+
+	if !all {
+		return true
+	}
+
+	return p.modelName != name
+}
+
+type osPolicy struct {
+	modelName string
+}
+
+func (p *osPolicy) CanRemove(st *state.State, snapst *snapstate.SnapState, all bool) bool {
+	name := snapst.InstanceName()
+	if name == "" {
+		// not installed, or something. What are you even trying to do.
+		return false
+	}
+
+	if boot.InUse(name, snapst.Current) {
+		return false
+	}
+
+	if !all {
+		return true
+	}
+
+	if snapst.Required {
+		return false
+	}
+
+	if name == "ubuntu-core" {
+		return true
+	}
+
+	return p.modelName != "" && !coreInUse(st)
+}
+
+type basePolicy struct{}
+
+func (basePolicy) CanRemove(st *state.State, snapst *snapstate.SnapState, all bool) bool {
+	name := snapst.InstanceName()
+	if name == "" {
+		// not installed, or something. What are you even trying to do.
+		return false
+	}
+
+	if boot.InUse(name, snapst.Current) {
+		return false
+	}
+
+	if !all {
+		return true
+	}
+
+	if snapst.Required {
+		return false
+	}
+
+	// here we use that bases can't be instantiated (InstanceName == SnapName always)
+	return !baseInUse(st, name)
+}
+
+func coreInUse(st *state.State) bool {
+	snapStates, err := snapstate.All(st)
+	if err != nil {
+		return err != state.ErrNoState
+	}
+	for name, snapst := range snapStates {
+		for _, si := range snapst.Sequence {
+			if snapInfo, err := snap.ReadInfo(name, si); err == nil {
+				if snapInfo.GetType() != snap.TypeApp || snapInfo.GetType() == snap.TypeSnapd {
+					continue
+				}
+				if snapInfo.Base == "" {
+					return true
+				}
+			}
+		}
+	}
+	return false
+}
+
+func baseInUse(st *state.State, baseName string) bool {
+	snapStates, err := snapstate.All(st)
+	if err != nil {
+		return false
```

Comment:
note how this and `coreInUse` react differently to errors that are not `ErrNoState`. I'm not sure I like that.